server/datastore/mysql/hosts.go

```diff
@@ -712,9 +712,19 @@ func (ds *Datastore) EnrollHost(ctx context.Context, osqueryHostID, nodeKey stri
 // LoadHostByNodeKey loads the whole host identified by the node key.
 // If the node key is invalid it returns a NotFoundError.
 func (ds *Datastore) LoadHostByNodeKey(ctx context.Context, nodeKey string) (*fleet.Host, error) {
-	sqlStatement := `SELECT * FROM hosts WHERE node_key = ?`
+	// LoadHostByNodeKey is one of the most used queries (every time a host checks in),
+	// thus prepare such query once.
+	ds.prepLoadHostQueryOnce.Do(func() {
+		var err error
+		//nolint ds.loadHostStmt is closed in Datastore.Close
+		ds.loadHostStmt, err = sqlx.PreparexContext(ctx, ds.reader, `SELECT * FROM hosts WHERE node_key = ?`)
+		if err != nil {
+			panic(fmt.Sprintf("prepare LoadHostByNodeKey statement: %s", err))
+		}
+	})
```

Comment:
Instead of the panic, you can declare `var err error` outside the func (before the call to the Once) and after the call check if err != nil and return that error.

Reply:
I did thought of that, but in case of `err != nil` one caller will get `err != nil` and the others will get `err == nil`

Reply:
Ah yeah you're right...

Reply:
Ah but the check could be `if err != nil || ds.loadHostStmt == nil` then you return an error (only one will get the actual real error, but others won't panic and can return an appropriate error message, like "failed to prepare load host statement"). Though thinking about it, if it did fail to prepare, we could fallback to running the unprepared query.

Reply:
I think we should move this into the constructor for the database ie in `func New(...) (*Datastore, error)`. There are several issues around handling errors in this function ie panic's, retries, etc. By moving it into the constructor, we can ensure it is only called once without using `sync.Once`.

Reply:
I agree that if possible, we should do without the `Once` and avoid races and more complex error handling/fallback. But based on another comment by Lucas on this PR, this seemed non-trivial to move there.

Reply:
Agree, but see https://github.com/fleetdm/fleet/pull/4128#discussion_r803074326

+1 to consider a more involved approach if we see that this change does improve performance.

Reply:
Hmm, I did not consider that. We could also move it to a separate function like `ds.PrepareStatements() error` that would get called after `ds.MigrateData()` in `main`. That way it would get called after the tables exist, and would still only get called once.

Reply:
+1, I was thinking similarly.